Upgrade PHPStan to version 1.0 - MAILPOET-3932

triple0t opened the pull request:
[MAILPOET-3932](https://mailpoet.atlassian.net/browse/MAILPOET-3932)

triple0t commented on tasks/phpstan/phpstan-baseline-fix-lib.php:
```diff
@@ -0,0 +1,20 @@
+$phpVersion = (int)getenv('ANALYSIS_PHP_VERSION') ?: PHP_VERSION_ID;
+$config['parameters']['phpVersion'] = $phpVersion;
+
+# PHPStan allows us to declare the currently reported list of errors as “the baseline” and cause it not being reported in subsequent runs.
+# PHPStan will throw violations only in new and changed code.
+# read more here: https://phpstan.org/user-guide/baseline
+# we need to load different baseline file based on the php version
```
I added a comment on what a baseline is here

triple0t commented on tasks/phpstan/phpstan.neon:
```diff
@@ -30,7 +30,8 @@ parameters:
     - '#Call to an undefined method MailPoetVendor\\Idiorm\\IdiormResultSet::set\(\)#'
     - '#Argument of an invalid type MailPoetVendor\\pQuery\\IQuery supplied for foreach, only iterables are supported#'
-    - '/(with|has) no (return )?typehint specified/' # exclude level 6 errors
+    - '/has no (return )?type specified/' # exclude level 6 errors
+    - '/has parameter (\$[_A-Z]{1}[_a-z]+)? with no type (specified)?/i' # exclude level 6 errors
```
I'm not sure if we still want to continue ignoring type errors, but I added these for backward compatibility

triple0t commented:
Premium version of the PR here [#523](https://github.com/mailpoet/mailpoet-premium/pull/523)

wxa commented:
@triple0t 

PHP 7.x and PHP 8.0 runs were green, but when I ran the 8.1 mode (`./do qa:phpstan --php-version=80100`) I got a syntax error in `tasks/phpstan/vendor/php-stubs/wordpress-stubs/wordpress-stubs.php`:
> Syntax error, unexpected T_READONLY, expecting T_STRING or '(' on line 98413

It's because WP (and its stubs) have a function named `readonly` which is a reserved keyword in PHP 8.1. [Here](https://github.com/php-stubs/wordpress-stubs/issues/20#issuecomment-970185024) it's advised to rename the stub, which I did in post-install (added a new commit), after which the syntax error was gone, but now I get 233 PHPStan errors as if a baseline file isn't included or something. 

Do you get anything similar? I'm using PHP 7.4 to run the command, by the way.

triple0t commented:
Hi @wxa 

Thanks for the comment.

I have fixed the PHP baseline errors for PHP 8.1

I got an error with this command on Mac OS:  `> sed -i -e 's#function readonly#function __readonly#' vendor/php-stubs/wordpress-stubs/wordpress-stubs.php`

```
sed: RE error: illegal byte sequence

Script sed -i -e 's#function readonly#function __readonly#' vendor/php-stubs/wordpress-stubs/wordpress-stubs.php handling the fixWPStubsForPHP81 event returned with error code 1
```
I refactored the function into a separate file _fix-WPStubs-for-PHP-8_1.php_

wxa commented on tasks/phpstan/fix-WPStubs-for-PHP-8_1.php:
```diff
@@ -0,0 +1,21 @@
+
+$search_term = 'function readonly';
+
+if (! str_contains($data, $search_term)) {
```
You are using a function only available in PHP 8, and we still haven't dropped PHP 7 support. Please use `strpos` or check PHP version before executing this code.

triple0t replied:
Noted @wxa

triple0t commented:
To maintain a clean git history, we are closing this PR in favor of [#3900](https://github.com/mailpoet/mailpoet/pull/3900)
